store/tracekvstore.go

```diff
@@ -0,0 +1,200 @@
+// KVStore implementation and a buffered writer.
+func NewTraceKVStore(parent sdk.KVStore, writer io.Writer, tc *TraceContext) *TraceKVStore {
+	return &TraceKVStore{parent: parent, writer: writer, context: tc}
+}
+
+// Get implements the KVStore interface. It traces a read operation and
+// delegates a Get call to the parent KVStore.
+func (tkv *TraceKVStore) Get(key []byte) []byte {
+	value := tkv.parent.Get(key)
+
+	writeOperation(tkv.writer, readOp, tkv.context, key, value)
+	return value
+}
+
+// Set implements the KVStore interface. It traces a write operation and
+// delegates the Set call to the parent KVStore.
+func (tkv *TraceKVStore) Set(key []byte, value []byte) {
+	writeOperation(tkv.writer, writeOp, tkv.context, key, value)
+	tkv.parent.Set(key, value)
+}
+
+// Delete implements the KVStore interface. It traces a write operation and
+// delegates the Delete call to the parent KVStore.
+func (tkv *TraceKVStore) Delete(key []byte) {
+	writeOperation(tkv.writer, deleteOp, tkv.context, key, nil)
+	tkv.parent.Delete(key)
+}
+
+// Has implements the KVStore interface. It delegates the Has call to the
+// parent KVStore.
+func (tkv *TraceKVStore) Has(key []byte) bool {
+	return tkv.parent.Has(key)
+}
+
+// Prefix implements the KVStore interface.
+func (tkv *TraceKVStore) Prefix(prefix []byte) KVStore {
+	return prefixStore{tkv, prefix}
+}
+
+// Iterator implements the KVStore interface. It delegates the Iterator call
+// the to the parent KVStore.
+func (tkv *TraceKVStore) Iterator(start, end []byte) sdk.Iterator {
+	return tkv.iterator(start, end, true)
+}
+
+// ReverseIterator implements the KVStore interface. It delegates the
+// ReverseIterator call the to the parent KVStore.
+func (tkv *TraceKVStore) ReverseIterator(start, end []byte) sdk.Iterator {
+	return tkv.iterator(start, end, false)
+}
+
+// iterator facilitates iteration over a KVStore. It delegates the necessary
+// calls to it's parent KVStore.
+func (tkv *TraceKVStore) iterator(start, end []byte, ascending bool) sdk.Iterator {
+	var parent sdk.Iterator
+
+	if ascending {
+		parent = tkv.parent.Iterator(start, end)
+	} else {
+		parent = tkv.parent.ReverseIterator(start, end)
+	}
+
+	return newTraceIterator(tkv.writer, parent, tkv.context)
+}
+
+type traceIterator struct {
+	parent     sdk.Iterator
+	key, value []byte
+	writer     io.Writer
+	context    *TraceContext
+}
+
+func newTraceIterator(w io.Writer, parent sdk.Iterator, tc *TraceContext) sdk.Iterator {
+	return &traceIterator{writer: w, parent: parent, context: tc}
+}
+
+// Domain implements the Iterator interface.
+func (ti *traceIterator) Domain() (start []byte, end []byte) {
+	return ti.parent.Domain()
+}
+
+// Valid implements the Iterator interface.
+func (ti *traceIterator) Valid() bool {
+	return ti.parent.Valid()
+}
+
+// Next implements the Iterator interface. It will write a traced operation
+// on a previously set key/value pair.
+func (ti *traceIterator) Next() {
```

Comment:
Hmm, I don't think this implementation is quite what we want - if I create an iterator, read one key, and never call `.Next()`, we won't log any operations - perhaps instead log key and value the first time `.Key()` is called, and reset a boolean on `.Next()`?